Enforce pre-commit in ArduPilot

Ryanf55 opened the pull request:
# Related issues

Fixes #24555. 
Depends on #24727.

# Purpose

Pre-commit provides basic guardrails for the repos, but it's not enforced. This PR adds it as a CI job. 

# How to test

```bash
CI_BUILD_TARGET=pre-commit-cleanliness ./Tools/scripts/build_ci.sh
```

# Future ideas

We could autofix and re-commit the changes in CI to save developer time, but one could see it as risky. 

Actions from Tridge's testing:
* We should add --allow-missing-config, this will support ArduPilot devs working on branches before 4.3. 
* Stop pre-commit from running when using subsystem split
* Need dev documentation on pre-commit

Ryanf55 commented:
Need to install pre-commit as part of the normal set of python packages. 
https://github.com/ArduPilot/ardupilot/actions/runs/5896623868/job/15994803894#step:4:142

khancyr reviewed (commented):
we should fix the files instead of excluding them.

tridge commented:
problem on older branches, this is on Copter-4.3:
![image](https://github.com/ArduPilot/ardupilot/assets/831867/f620a868-6e71-4e01-a410-6033e92b418d)

Ryanf55 commented:
> problem on older branches, this is on Copter-4.3: ![image](https://user-images.githubusercontent.com/831867/262188108-f620a868-6e71-4e01-a410-6033e92b418d.png)

Looks like we might be able to force an upgrade of `isort` to fix this: https://github.com/home-assistant/core/issues/86892
I'll give it a whirl.

Ryanf55 commented on .pre-commit-config.yaml:
```diff
@@ -24,16 +24,45 @@ repos:
+            exclude:
+              (?x)^(
+                Tools/bootloaders/CubeOrange-bdshot_bl.hex|
+                libraries/AP_Baro/AP_Baro_ICP101XX.cpp |
```
See the other PR, fixed the current errors.

Ryanf55 commented on .pre-commit-config.yaml:
```diff
@@ -24,16 +24,45 @@ repos:
+            exclude: |
+              (?x)^(
+                Tools/autotest/template/scrimmage.xml |
+                wscript
```
All the files are now fixed, the pre-commit has no new excludes.

Ryanf55 commented:
Rebased and fixed the regression:
```
ryan@B650-970:~/Dev/ros2_ws/src/ardupilot$ pre-commit run --all-files
Check line ending character (LF).........................................Failed
- hook id: mixed-line-ending
- exit code: 1
- files were modified by this hook

libraries/SITL/SITL_Airspeed.cpp: fixed mixed line endings
```

peterbarker commented on .github/workflows/test_scripts.yml:
```diff
@@ -19,6 +19,7 @@ jobs:
             python-cleanliness,
             astyle-cleanliness,
             validate_board_list,
+            pre-commit-cleanliness,
```
Would need an entry in `Tools/scripts/builds_ci.sh` IIUC

Ryanf55 replied:
Fixed, accidentally dropped it during the rebase. It was originally there on this commit. 
https://github.com/ArduPilot/ardupilot/commit/48cc7b110787984eb86527cd81e4f0028b98ee61#diff-1600157c5dce475648f144b0f9b010395f5a9d5ad66a3fe9e56a3508cd930c67

Ryanf55 commented:
Here's a demo for the CI script with a good PR: 
```
ryan@B650-970:~/Dev/ros2_ws/src/ardupilot$ CI_BUILD_TARGET=pre-commit-cleanliness  ./Tools/scripts/build_ci.sh
...
+ echo 'Checking pre-commit code cleanliness'
Checking pre-commit code cleanliness
+ pre-commit install
pre-commit installed at .git/hooks/pre-commit
+ pre-commit run --all-files --verbose --show-diff-on-failure
Check line ending character (LF).........................................Passed
- hook id: mixed-line-ending
- duration: 0.06s
check for added large files..............................................Passed
- hook id: check-added-large-files
- duration: 0.06s
check that executables have shebangs.....................................Passed
- hook id: check-executables-have-shebangs
- duration: 0.04s
check that scripts with shebangs are executable..........................Passed
- hook id: check-shebang-scripts-are-executable
- duration: 0.06s
check for merge conflicts................................................Passed
- hook id: check-merge-conflict
- duration: 0.05s
check xml................................................................Passed
- hook id: check-xml
- duration: 0.04s
check yaml...............................................................Passed
- hook id: check-yaml
- duration: 0.04s
Formats XML files........................................................Passed
- hook id: format-xmllint
- duration: 0.02s

XML check for libraries/AP_DDS/dds_xrce_profile.xml: OK

black....................................................................Passed
- hook id: black
- duration: 0.07s

All done! ✨ 🍰 ✨
2 files left unchanged.

+ continue
+ echo build OK
build OK
+ exit 0
```

Now, if we accidentally make a cpp file executable, you get this in the CI logs:
```
+ pre-commit run --all-files --verbose --show-diff-on-failure
Check line ending character (LF).........................................Passed
- hook id: mixed-line-ending
- duration: 0.06s
check for added large files..............................................Passed
- hook id: check-added-large-files
- duration: 0.06s
check that executables have shebangs.....................................Failed
- hook id: check-executables-have-shebangs
- duration: 0.04s
- exit code: 1

ArduCopter/AP_ExternalControl_Copter.cpp: marked executable but has no (or invalid) shebang!
  If it isn't supposed to be executable, try: `chmod -x ArduCopter/AP_ExternalControl_Copter.cpp`
  If on Windows, you may also need to: `git add --chmod=-x ArduCopter/AP_ExternalControl_Copter.cpp`
  If it is supposed to be executable, double-check its shebang.

check that scripts with shebangs are executable..........................Passed
- hook id: check-shebang-scripts-are-executable
- duration: 0.06s
check for merge conflicts................................................Passed
- hook id: check-merge-conflict
- duration: 0.05s
check xml................................................................Passed
- hook id: check-xml
- duration: 0.03s
check yaml...............................................................Passed
- hook id: check-yaml
- duration: 0.04s
Formats XML files........................................................Passed
- hook id: format-xmllint
- duration: 0.02s

XML check for libraries/AP_DDS/dds_xrce_profile.xml: OK

black....................................................................Passed
- hook id: black
- duration: 0.07s

All done! ✨ 🍰 ✨
2 files left unchanged.
```

Ryanf55 commented:
@khancyr Can you create a new tag/release for the dev dockerfiles? `v0.1.3` ,which is being used above, results in the failed build because it's missing xmllint.

khancyr commented:
my bad, I haven't seen it wasn't in the release. I have force push a rebuild with the latest version

MichelleRos commented:
Is there an option to disable this pre-vcommit though? I *really* wouldn't want to be forced to obey specific rules when I'm just doing test development in my own fork.

That would really slow down development.

Ryanf55 commented:
> Is there an option to disable this pre-vcommit though? I _really_ wouldn't want to be forced to obey specific rules when I'm just doing test development in my own fork.
> 
> That would really slow down development.

Yep! `git commit --no-verify` disables it when you commit. 

That said, do you have examples of this slowing down development and specifics on which hooks or workflows have affected you? We only run a bare minimum set of hooks right now.

khancyr commented on Tools/scripts/build_ci.sh:
```diff
@@ -424,6 +424,13 @@ for t in $CI_BUILD_TARGET; do
 
+    if [ "$t" == "pre-commit-cleanliness" ]; then
+        echo "Checking pre-commit code cleanliness"
+        pre-commit install
```
@Ryanf55 that means that after a first run precommit will be enable on git commit. Probably not the best move. Just doing the pre-commit run is enough and don't mess the git config for now (I still want pre-commit)

Ryanf55 replied:
Are you saying that local developers run build_ci with specific arguments to hit this code block?

peterbarker replied:
@Ryanf55 I think Pierre is saying we shouldn't be making this permanent change to their setup in here.

We can do it as part of the machine setup in the github workflows instead, yes?

Ryanf55 replied:
yes, but I thought it would be ok because this is a CI script. I'm happy to move it if you think we need to worry about devs running this check and then being surprised pre-commit is now enabled.

peterbarker replied:
Yes, please move it.

Ryanf55 replied:
Done! Let's see how it goes...

khancyr replied:
@Ryanf55 normally there is not issue to use `pre-commit run --all-files` on CI, install isn't need for CI

magicrub commented:
@Ryanf55 This is tagged MergeOnCiPass, and it's passing, but it has a merge conflict

Ryanf55 commented:
> @Ryanf55 This is tagged MergeOnCiPass, and it's passing, but it has a merge conflict

I never got an answer on Pierre's comment. I don't have a path forward on this without feedback on the raised concern.

magicrub commented:
@khancyr ping!

khancyr commented on .github/workflows/test_scripts.yml:
```diff
@@ -19,12 +19,18 @@ jobs:
+      - name: install pre-commit
+        shell: bash
+        run: |
+          pre-commit install
```
why need this ? we don't run git command on CI, so the install isn't need normally, or I miss something ?

khancyr replied:
are you sure that shouldn't be `pip install pre-commit `instead ?

khancyr replied:
OK i have checked and we got pre-commit into the docker image already so we are fine. the pre-commit install is just to install the git hook so that isn't need on CI

Ryanf55 replied:
You are conflating installation of the pre-commit tool in the environment and installation of the git hooks into the repo. 
https://pre-commit.com/#3-install-the-git-hook-scripts

We need both at some point before calling `pre-commit run` in ArduPilot.

khancyr replied:
Nope you don't need to install the git hook to use precommit.

khancyr reviewed (changes requested):
something is wrong with this commit as it moves the submodules !

peterbarker commented:
70 failing tests, un-addressed review comments, many conflicts.... and commits into the submodules?!

It looks like we might have been close to merging this one... but the ball's in your court here, @Ryanf55 

Are you still wanting to pursue this one?

Ryanf55 commented:
> 70 failing tests, un-addressed review comments, many conflicts.... and commits into the submodules?!
> 
> It looks like we might have been close to merging this one... but the ball's in your court here, @Ryanf55
> 
> Are you still wanting to pursue this one?

I gave up - someone else can do this.
